db/system_keyspace.cc

```diff
@@ -258,6 +258,7 @@ schema_ptr system_keyspace::topology() {
             .with_column("ignore_msb", int32_type)
             .with_column("supported_features", set_type_impl::get_instance(utf8_type, true))
             .with_column("new_cdc_generation_data_uuid", uuid_type, column_kind::static_column)
+            .with_column("in_progress_cdc_generation_id", uuid_type, column_kind::static_column)
```

Comment:
> Topology snapshots contain only mutation of current CDC generation data but don't
contain any previous or future generations. If new a generation of data is being
broadcasted but hasn't been entirely applied yet, the applied part won't be sent
in a snapshot. In this scenario, new or delayed nodes can never get the applied part.

So how this can happen? If we broadcast new CDC generation it means we either bootstrap a node or execute "cdc generation recalculation" command. In both cases we cannot join another node until the operation finishes (either succeeds or fails).

Well, right now a new node joins group0 and adds itself to tho topology and that may happen while another operation is in progress, but we are changing it. The coordinator will add nodes to group0 only after previous operation changes, so I think all this complexity will not be needed.

Reply:
> so I think all this complexity will not be needed.

We still need to be able to transfer a snapshot if there is a slow follower and a snapshot is taken while cdc generation is transferred. IMO it just shows that we should use storage proxy for that.

Reply:
> We still need to be able to transfer a snapshot if there is a slow follower and a snapshot is taken while cdc generation is transferred.

Yup. Besides, relying on nodes not joining group 0 during bootstrap is not quite robust - what if we allow concurrent bootstraps in the future?

> IMO it just shows that we should use storage proxy for that.

I disagree.

Reply:
> > We still need to be able to transfer a snapshot if there is a slow follower and a snapshot is taken while cdc generation is transferred.
> 
> Yup. Besides, relying on nodes not joining group 0 during bootstrap is not quite robust - what if we allow concurrent bootstraps in the future?

Concurrent bootstrap will join `n` nodes simultaneously, not add other nodes during ongoing operation.

> 
> > IMO it just shows that we should use storage proxy for that.
> 
> I disagree.

Considering that according to the logic I described streaming will not be needed what is your justification for a more complicated solution?

Reply:
And SP will do it more efficiently as well as it knows not to send a mutation to each node in another DC.

Reply:
>  a more complicated solution?

But it's your solution which is more complicated, not mine. I suggest we put everything in group 0, you suggest to put part of cdc outside group 0 and have another part that lives inside group 0 depends on the outside part, and you claim this is less complicated?

Reply:
> Considering that according to the logic I described streaming will not be needed

How so?

Reply:
> > Considering that according to the logic I described streaming will not be needed
> 
> How so?

Because no node will join while another node is joining already. It means that the partial data will never have to be streamed.

Reply:
> > a more complicated solution?
> 
> But it's your solution which is more complicated, not mine. I suggest we put everything in group 0, you suggest to put part of cdc outside group 0 and have another part that lives inside group 0 depends on the outside part, and you claim this is less complicated?

Yes. You need to write the data to all node. I propose to use existing infra (mutation_write_handlers) to do so. Scylla is all about writing data to nodes.

Reply:
>  It means that the partial data will never have to be streamed.

This "partial data" becomes "full data" once we write all of it successfully and we commit the CDC generation ID using `topology_change`. Then it will have to be streamed.

Reply:
> I propose to use existing infra (mutation_write_handlers) to do so. 

you're contradicting yourself from the other thread:

>    I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.

Reply:
> > It means that the partial data will never have to be streamed.
> 
> This "partial data" becomes "full data" once we write all of it successfully and we commit the CDC generation ID using `topology_change`. Then it will have to be streamed.

After that point it is transferred as part of the topology snapshot. I do not propose to change that.

Reply:
Let's discuss this on todays daily

Reply:
> > I propose to use existing infra (mutation_write_handlers) to do so.
> 
> you're contradicting yourself from the other thread:
> 
> > I propose to write the code similar to what we have for batches in storage_proxy now. It will construct specially crafted mutation_write_handler with all the targets that we want to write into.

Those two statement are literally identical.

Reply:
Summarized in https://github.com/scylladb/scylladb/pull/13962#issuecomment-1609112962